### ISSUE TITLE
depending on call sequence, newsplit() can break bind strings

### ISSUE THREAD

**Issue by thommey** [eggheads/eggdrop#387]
From #365 
> Further testing reveals two things:
 (1) mode bind for - modes triggers BEFORE raw mode binds (whereas + binds trigger AFTER the raw mode bind)
 (2) there's a SECOND 0x00 in the - mode raw: "nick(0x00)ident@host(0x00)#channel (+/-)mode target"

Found by @maimizuno

Under some circumstances it can happen that bind mode trigger before bind raw - in which case bind mode uses newsplit() on the user and thereby destroys the string for the raw bind call. we should at least rule it out in that case. It seemed to depend on whether it's -o or +o

**Comment by thommey** [comment]
Unfortunately I was unable to reproduce this again. No matter what I do, I couldn't get Eggdrop to ever call the mode bind before the raw bind, which is correct behavior. I will leave this open because I saw it happening and I know I was able to reproduce it at some point (at least the reversed order), but not anymore.

**Comment by ghost** [comment]
Most bare-bones example I can produce:

\<Domino> do tcl bind raw - mode bind:raw:mode ; proc bind:raw:mode args { putquick "PRIVMSG #Channel :\[[string toupper [lindex [info level [info level]] 0]]\] ARGS($args)" ; return 0 }
-Dorito- OK.

\<Domino> do tcl bind mode - * bind:mode:mode ; proc bind:mode:mode args { putquick "PRIVMSG #Channel :\[[string toupper [lindex [info level [info level]] 0]]\] ARGS($args)" ; return 0 }
-Dorito- OK.

\* Domino sets mode: +v Domino
\<Dorito> [BIND:RAW:MODE] ARGS(Domino!Domino@Coder MODE {#Channel +v Domino})
\<Dorito> [BIND:MODE:MODE] ARGS(Domino Domino@Coder Domino #Channel +v Domino)
\* Domino sets mode: -v Domino
\<Dorito> [BIND:MODE:MODE] ARGS(Domino Domino@Coder Domino #Channel -v Domino)
\<Dorito> [BIND:RAW:MODE] ARGS(Domino

As you can see, the following problems occur:

(1) +mode: RAW before MODE
(2) -mode: MODE before RAW
(3) -mode: RAW args: the space between nick and host, &, space between host and channel, both are replaced with 0x00 (as previously reported)

This is a non-enhanced raw Eggdrop bot, not enhanced with my standard code packages.

**Comment by michaelortmann** [comment]
its not about -mode or +mode, but, as in #354 about the order of execution. the order of internal gotmode() vs. external raw proc toggles. gotmode() is changing args. we should analyze/fix the order of execution.

**Comment by ghost** [comment]
```
Discussion in channel between me, @vanosg and @michaelortmann revealed these issues 
with this and related PRs: (#354, # 365, & #387). To combine them into one place, 
we've concluded this so far:

(I) any mode changes to a channel, regardless of number of targets:
    (A) any mode change will work properly: the RAW bind will trigger
        Ami-Laptop!Mercury@ShabonSpray.ShineAquaIllusion.Org MODE {#Testing +i }
    (B) then the internal MODE bind will trigger
        Ami-Laptop Mercury@ShabonSpray.ShineAquaIllusion.Org Domino #Testing +i {}

(II) the next mode change will be processed differently:
    (A) the INTERNAL bind will trigger first
        Ami-Laptop Mercury@ShabonSpray.ShineAquaIllusion.Org Domino #Testing -i {}
    (B) then the RAW bind will trigger, but two different things happen:
        (1) Ami-Laptop*[0x00]*Mercury@ShabonSpray.ShineAquaIllusion.Org Domino #Testing -i {}
            the space between the nick and host is replaced by a null char (0x00)
        (2) the order has changed from raw / internal, to internal/raw
        (3) the raw bind (II/B) comed out delayed by nearly 2/3 of a second

(III) there's an ambiguity, therefore, as to what order the binds SHOULD trigger:
    (A) raw, then internal
    (B) internal then bind

(IV) RAW bind: return value (return 1) indicates a short-circuit of bind processing 
     (per the docs). As such, it is believed (for now), that the RAW bind should come first;
     logically, having a MODE bind trigger, then the RAW bind (which can kill any further bind
     triggers) seems backward.


(v) issues to solve (as of now):
    (A) determine the order; need to confer with the OG coders (Geo, thommey, et cetera)
    (B) determine why this "toggle" exists
    (C) determine if the 0x00 substitution it related to the above, or, another issue altogether
```

**Comment by michaelortmann** [comment]
the mysterious toggle seems to be happening here, i will analyze later, for now i cant say what side effects removal may have, it is there for a reason i guess! but if you wanna test:
in src/tclhash.c remove the 6 lines in function check_tcl_bind():

  if (tm_p && tm_p->next) {
    tm = tm_p->next;            /* Move mask to front of bind's mask list. */
    tm_p->next = tm->next;      /* Unlink mask from list. */
    tm->next = tl->first;       /* Readd mask to front of list. */
    tl->first = tm;
  }

this block looks like an optimization with no other function, so that recently used binds are scanned first the next time the bind list is scanned. if that's the case, we should simply remove that block.

**Comment by ghost** [comment]
Testing of that edit causes the following results:

(1) binds are always processed RAW, then internal binds, LIFO order
(2) RAW return 1 still executes the RAW bind but stops all other binds
(3) the 0x00 issue is now gone

I also tested with multiple mode binds stacked; all modes fired in LIFO order; the new code's RAW bind came first; the old code's RAW order displayed the toggle as noted earlier.

I see no [unexpected] side-effects.

**Comment by vanosg** [comment]
Closed by #566
